File: src/SubtitleMatcher.ts
import { AssertionError } from "assert";
import fs from "fs-extra";
import path from "path";
import { zip } from "./misc";


export class SubtitleMatcher {
	private readonly regexp: RegExp;

	constructor(private vid_ext: string, private sub_ext: string, private folder: string) { 
		this.regexp = new RegExp(`${this.vid_ext}\$`)
	}

	public match(): void {
		let items = fs.readdirSync(this.folder)
			.map(basepath => path.join(this.folder, basepath));
		let videos = items.filter(i => i.endsWith(this.vid_ext));
		let subtitles = items.filter(i => i.endsWith(this.sub_ext));

		let sorter = Intl.Collator(undefined, { numeric: true }).compare;
		videos.sort(sorter);
		subtitles.sort(sorter);

		if (videos.length != subtitles.length) {
			throw new AssertionError({
				message: `# of videos(${videos.length}) doesn't match # of subtitles(${subtitles.length})`
			});
		}

		zip(videos, subtitles)
			.forEach(([v, s]) => {
				const new_sub_name = this.rename(v)
				fs.rename(s, new_sub_name)
					.then(() => console.log(`${path.basename(s)} -> ${path.basename(new_sub_name)}`))
			});
	}

	/**
	 * It transforms its extension to `smi | srt` without changing the basename.
	 * @param src video file name
	 */
	private rename(src: string): string {
		const new_sub_name = src.replace(this.regexp, this.sub_ext);
		return new_sub_name;
	}
}

File: subtitle_matcher.py
#!/usr/bin/env python3
from argparse import ArgumentParser
import os
from pathlib import Path
from re import compile, split
from typing import Literal, Optional

import tabulate


class SubtitleMatcher:
	def __init__(self, video_ext: str, folder: str, suffix: Optional[str] = None) -> None:
		self.video_ext = video_ext
		self.folder = folder
		self.regexp = compile(f"{self.video_ext}$")

		self.suffix = suffix

	def match(self):
		videos, subtitles = self._get_sorted_file_list()

		line_list = []
		for v, s in zip(videos, subtitles):
			subtitle_ext = s.split('.')[-1]
			new_name = self._get_new_name(v, subtitle_ext)  # type: ignore
			os.rename(s, new_name)

			line_list.append( (os.path.basename(s) , os.path.basename(new_name)))
		print(tabulate.tabulate(line_list, headers=["From", "To"], tablefmt="fancy_grid"))

	def _get_sorted_file_list(self):
		items = os.listdir(self.folder)
		items = [ str(Path(self.folder).joinpath(it)) for it in items ]

		videos = [it for it in items if it.endswith(self.video_ext)]
		subtitles = [it for it in items if it.endswith(".smi") or it.endswith(".srt")]

		videos.sort(key=NumericalSorter())
		subtitles.sort(key=NumericalSorter())

		if(len(videos) != len(subtitles)):
			raise RuntimeError("len(videos) != len(subtitles)")

		return videos, subtitles
	
	def _get_new_name(self, src: str, subtitle_ext: Literal["smi", "srt"]) -> str:
		if self.suffix:
			return self.regexp.sub(f"{self.suffix}.{subtitle_ext}", src)
		return self.regexp.sub(subtitle_ext, src)


class NumericalSorter:
	def __call__(self, s: str):
		return [self._try_int(i) for i in split(r"(\d+)", s)]
	
	@staticmethod
	def _try_int(n: str):
		if n.isdecimal():
			return int(n)
		return n




def main(args):
	s = SubtitleMatcher(args.video_ext, args.folder, args.suffix)
	s.match()


if __name__ == "__main__":
	parser = ArgumentParser()
	parser.add_argument("-f", "--folder", help="folder containing movies", default="./")
	parser.add_argument("-v", "--video_ext", help="video extension", choices=["mkv", "mp4", "avi"], required=True)
	parser.add_argument("--suffix", help="optional suffix", default=None)

	main(parser.parse_args())


